src/shared/wizards/wizardForm.ts

```diff
@@ -3,157 +3,533 @@
+ * some property called `foo` that has a type of `string`, then the callback must return a `Prompter<string>`
+ * if binding to `foo`.
+ *
+ * These callbacks can declare dependencies on various properties of the state to ensure that they are
+ * defined at runtime by using {@link ContextOptions.dependencies} or by manually setting the `TDep` template
+ * type to a subtype of an arary of {@link Pathable Pathables}.
+ */
+export type PrompterProvider<TState, TProp, TDep> = (
+    state: StateWithCache<BoundState<TState, TDep>, TProp>
+) => Prompter<TProp>
+
+/**
+ * Resolves a state with the associated dependency bindings.
+ *
+ * This uses {@link FindPartialTuple} to recursively set props provided by {@link ExtractPaths}.
+ *
+ * Example:
+ * ```ts
+ * // `state` as an object is a runtime construct, so we use the type instead
+ * // if you had the actual state, you wouldn't need these types!
+ * type State = {
+ *    foo?: {
+ *       bar: string
+ *       qaz: string
+ *    }
+ *    baz: string
+ * }
+ * const bindings = [
+ *   { path: ['baz'] as const },
+ *   { path: ['foo', 'bar'] as const },
+ * ]
+ *
+ * type StateWithBindings = BoundState<State, typeof Bindings>
+ * // Equivalent to
+ * type StateWithBindings = {
+ *    foo: {
+ *       bar: string
+ *       qaz?: string
+ *    }
+ *    baz: string
+ * }
+ * ```
+ */
+export type BoundState<TState, Bindings> = FindPartialTuple<TState, ExtractPaths<Bindings>>
+
+/**
+ * Sets a 'default' value to a property of the state.
+ *
+ * This is similar to {@link PrompterProvider} but is intended to be more 'ephemeral' and not
+ * necessarily a strict binding. A prompter (if shown) would override any defaults set by this function.
+ */
+type DefaultFunction<TState, TProp, TDep> = (state: BoundState<TState, TDep>) => TProp
+
+/**
+ * Changes all properties enumerated by `K` to be required while leaving everything else
+ * to be optional. For example, consider `K` as ['foo', 'bar'] for a `T` of:
+ * ```ts
+ * type State = {
+ *    foo?: {
+ *       bar: string
+ *       qaz: string
+ *    }
+ *    baz: string
+ * }
+ * ```
+ * Would become:
+ * ```ts
+ * type PartialState = FindPartialState<State, ['foo', 'bar']>
+ * // Equivalent to
+ * type PartialState = {
+ *    foo: {
+ *       bar: string
+ *       qaz?: string
+ *    }
+ *    baz?: string
+ * }
+ * ```
+ * `M` is used to describe when there is a matched key. Matched keys should return their type,
+ * while non-matches are considered to be unassigned, and thus are potentially undefined.
+ */
+type FindPartialTuple<T, K extends string[], M extends boolean = false> = K['length'] extends 0
+    ? M extends true
+        ? T
+        : RecursivePartial<T>
+    : {
+          [P in keyof T as P & K[0]]-?: FindPartialTuple<T[P], ReduceTuple<K>, true>
+      } &
+          {
+              [P in keyof T as Exclude<P, K[0]>]+?: FindPartialTuple<T[P], ReduceTuple<K>, false>
```

Comment:
whoa - I have no idea what this is doing!

Reply:
Yeah there's a lot of specialized syntax here that's unlike most languages:
* `K['length'] extends 0` is a conditional type to check for an empty tuple
* `M extends true` is there in case we ended on a non-primitive type, in which case we should either pass-through or make it all optional.
* There are two uses of mapped types which have this basic shape: `{ [P in keyof T]: T[P] }` (identity mapping)
  * Mapped types are pretty much what they sound like: they map one type to another. Types in the square brackets are the key, and the right side is the desired type.
* `P in keyof T as P & K[0]` is essentially saying `P` is all keys of `T` that are also the first element of the `K` tuple. The `as` keyword is specifying a mapped key, e.g. `P'(P) = P & K[0]`
* `-?` is a modifier that makes the new type 'non-nullable' (not undefined or null)
* `P in keyof T as Exclude<P, K[0]>` is all keys of `T` that are _not_ the first element of `K`
* `+?` is a modifier that makes the new type 'optional' (so it can be undefined)
* `FindPartialTuple` is the recursive part of the alias. `T[P]` is the type associated with the property key `P` in `T`.
* `ReduceTuple` can be thought of as shifting the tuple, so just removing the first element.

Essentially the alias provides a way to make arbitrary properties of an interface/type required while making everything else optional. In general, it's best to think of Typescript aliases as math functions/definitions considering there's no way to write aliases with an imperative style.